### PR TITLE
fix(app): resume phone mic after other-app audio on iPadOS (#4706)

### LINKED ISSUE #4706: Recording stops / stuck when other audio plays — Omi does not auto-resume (iPadOS)
This issue is linked to our feedback platform. For more details and updates, please visit [this link](https://omi.featurebase.app/p/recording-stops-stuck-when-other-audio-plays-omi-does-not).

---

Recording stops / stuck when other audio plays — Omi does not auto-resume (iPadOS)
When Omi is recording on iPad and another app starts playing audio (e.g. a YouTube video in Safari), even without the app utilising microphone access itself, Omi loses microphone input and the recording session becomes stuck. Omi appears to still show as “recording” but no audio is captured and it does not auto-resume when the other audio stops.

Steps to reproduce:

Open Omi app on iPad (use Stage Manager/Windowed-Apps/multi-app if relevant).

Start recording (with any transcription provider, as it was observed in both Omi cloud & On-Device iOS STT).

While recording, open Safari and play a YouTube video (or any other app that plays audio).

Observe: Omi stops receiving mic input; the UI may still show recording, but transcript is empty / stuck.

Stop video. Omi does not automatically resume capturing. User must end session and start a new one.

Observed: microphone input is interrupted, recording stuck, no auto-resume.

Expected: Omi should detect AVAudioSession interruption, pause gracefully, and either auto-resume when mic returns or show explicit “Recording paused due to system audio” with a one-tap resume.

Environment / important details:

Device: iPad Pro (12.9-inch) (5th generation)

iPadOS version: 26.3

Omi app version: 1.0.523

Mode: Windowed Apps

Transcription mode tested: Omi cloud & On-Device iOS Speech

Voice Isolation / Live Listen: OFF (I tested with them off)

Repro rate: every time

### PR DESCRIPTION
**Pull request by aryanorastar**
## What changed and why

On iPadOS (Stage Manager / windowed apps), playing YouTube (or other audio) while Omi phone-mic records can stop mic frames without an AVAudioSession interruption. The UI stayed on “Listening”, Dart stall timers may not fire while suspended, and `appBecameActive` only resumed if already `.interrupted`. Fixes #4706.

- **Native:** on `appBecameActive` while `.running`, rebuild if the engine is dead; on `interruptionEnded`, always `attemptResume()` once (other-app audio often ends without `shouldResume`).
- **Dart:** `probeStallAfterForeground()` soft-rearms the stall clock from home lifecycle resume (does **not** immediately stop→start — native owns dead-engine rebuild; the periodic watchdog escalates only if frames still don’t arrive).
- **UI:** any `RecordingState.interrupted` shows Paused (not only call-active).

## Product invariants affected

none

## How it was verified

```bash
cd app && flutter test \
  test/unit/native_mic_recorder_service_test.dart \
  test/unit/mic_arbiter_test.dart \
  test/unit/voice_recorder_mic_contention_test.dart \
  test/widgets/transcription_paused_warning_test.dart
```

Could not reproduce on a physical iPad in this environment. Native dead-engine rebuild is covered by the existing PhoneMic control path + unit tests for the Dart timer-suspension gap.

## Tests

- `app/test/unit/native_mic_recorder_service_test.dart` — soft-rearm / watchdog foreground-stall cases
- `app/test/widgets/transcription_paused_warning_test.dart` — non-call interrupted → Paused + paused affordance
- mic test doubles stub `probeStallAfterForeground`

## Failure class (fixes)

Failure-Class: none

## Scoped cleanups (optional)

n/a

### CODE REVIEW

**Comment by aryanorastar** [comment]
**CI note:** Backend Hermetic Merge Gate red is not from this PR’s app changes.

Same flake as #11144: `Detect Hermetic Backend Scope` hit the **2m job timeout** and was cancelled → Merge Gate fail-closed on `SCOPE_RESULT=cancelled` (E2E correctly skipped). No admin rights to `gh run rerun`; empty commit retriggers checks.

**Comment by aryanorastar** [comment]
## Follow-up @6d695a20 (cubic + Formatting)

Addressed all three cubic findings:

1. **P2 dual recovery** — foreground no longer Dart stop→starts; soft-rearms stall clock only. Native owns dead-engine rebuild.
2. **P2 false stall** — wall-clock silence from suspension alone no longer escalates; watchdog still catches real post-resume silence.
3. **P3 UI test** — widget coverage for non-call interrupted → Paused.

Also re-ran `dart format --line-length 120` on the touched Dart files (Formatting CI was red on the previous push).

File: app/lib/services/capture/capture_controller.dart [hunk 1]
```diff
@@ -254,6 +254,14 @@ class CaptureController extends ChangeNotifier
     }
   }
 
+  /// Foreground return hook for phone-mic capture (#4706). Native also rebuilds
+  /// a dead engine on `didBecomeActive`; this covers the Dart timer-suspension
+  /// case where frames stopped without an interruption notification.
+  void onAppResumed() {
+    if (_activeSource is! PhoneMicSource && !_phoneMicBatchActive) return;
+    ServiceManager.instance().phoneMic.probeStallAfterForeground();
```
[the remark this answers is not in the document]
Comment (aryanorastar) [reply app/lib/services/capture/capture_controller.dart:265]: Addressed in 6d695a20 — see PR comment for details.

[the remark this answers is not in the document]
Comment (aryanorastar) [reply app/lib/services/capture/capture_controller.dart:265]: Addressed in 6d695a20 — see PR comment for details.

File: app/lib/pages/conversations/widgets/processing_capture.dart [hunk 2]
```diff
@@ -346,37 +350,39 @@ class _ConversationCaptureWidgetState extends State<ConversationCaptureWidget> {
...
             provider.recordingState == RecordingState.initialising ||
             provider.recordingState == RecordingState.interrupted ||
             _isPhoneMicPaused);
 
     // Determine pause state based on recording type.
-    // Call-active interruption is treated as paused for button/dot display.
-    bool isCallInterrupted = provider.recordingState == RecordingState.interrupted && provider.isCallActive;
+    // Any audio-session interruption (call, other-app audio, system alert) is
+    // treated as paused so the UI does not claim "Listening" while mute (#4706).
+    bool isAudioInterrupted = provider.recordingState == RecordingState.interrupted;
```
[the remark this answers is not in the document]
Comment (aryanorastar) [reply app/lib/pages/conversations/widgets/processing_capture.dart:359]: Addressed in 6d695a20 — see PR comment for details.

[the remark this answers is not in the document]
Comment (aryanorastar) [reply app/lib/pages/conversations/widgets/processing_capture.dart:359]: Strengthened in the follow-up push: the #4706 widget test now asserts **Paused** text, the orange paused status dot, and the play (resume) affordance for `isCallActive == false`.

Also fixed Dart Analyze: test doubles implementing `IMicRecorderService` now stub `probeStallAfterForeground()` (that was the red CI).

**Comment by aryanorastar** [comment]
## CI fix

**Dart Analyze & Tests** failed because `FakeMic` / `_ContendedMic` did not implement the new `IMicRecorderService.probeStallAfterForeground()` — compile-time, never baselined.

Fixed + pushed; widget test also asserts paused affordance (orange status dot + play icon) per cubic P3.

**Comment by Git-on-my-level** [comment]
Thanks for this — clean, well-scoped fix for a real iPadOS edge case (#4706: other-app audio silently killing the phone-mic engine without an `AVAudioSession` interruption, leaving the UI stuck on "Listening" while mute).

A few notes from review:

- The split of responsibility reads correctly: native `appBecameActive` owns dead-engine rebuild, while `probeStallAfterForeground()` on the Dart side only **soft-rearms** the stall clock instead of forcing a stop→start. Avoiding that Dart-side escalation racing the native rebuild is the right call, and the test in `native_mic_recorder_service_test.dart` covers both the soft-rearm and the genuine post-resume silence path.
- Generalizing the interrupted→Paused UI display (not only `isCallActive`) is a good correctness improvement; the widget test asserts the affordance.
- The `interruptionEnded` change to always `attemptResume()` once is reasonable given that other-app audio frequently ends without `shouldResume`. The ticker keeps probing if bring-up is still blocked, so the failure path stays safe.

I see the prior cubic-dev-ai findings (dual-recovery race, false-stall on suspended timers, missing UI test) were all addressed in the follow-up commit, and CI is green.

One thing a maintainer should confirm on a physical iPad before merge: the `appBecameActive` + `.running` + dead-engine rebuild path can't be exercised by CI, so a quick on-device smoke (Omi recording → launch YouTube/Stage Manager → return to Omi → verify frames resume) would close the loop. No code changes requested from this review.

Leaving for a human maintainer to verify the on-device behavior and merge.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Comment by aryanorastar** [comment]
@Git-on-my-level Thanks for the positive signal.

Agreed on the iPad smoke gap — no physical iPad in this environment, so that path stays an honest verification hole for a maintainer with device access (Omi recording → YouTube/Stage Manager → return → frames resume).

No further author code changes pending on this head; CI green @b82fb8ca including the soft-rearm + widget-affordance follow-ups.

**Comment by aryanorastar** [comment]
@Git-on-my-level need human response

Head `@b82fb8ca` is fully green (Dart Analyze, Android Compile, Preflight). Cubic findings addressed; your positive-signal pass found no code changes. Remaining gap is honest: physical iPad smoke (Omi recording → YouTube/Stage Manager → return → frames resume).

Merge still blocked solely on write-access APPROVE. Could a human maintainer with device access smoke + APPROVE/merge when you have a moment? No further author work pending.

**Review by kodjima33 — APPROVED** [review]
Rebuilds the dead phone-mic engine on foreground and always retries on interruptionEnded; covers the iPadOS other-app-audio stall.
